mkosi/vmspawn.py

```diff
@@ -0,0 +1,165 @@
+import contextlib
+import os
+import subprocess
+import sys
+from pathlib import Path
+
+from mkosi.config import (
+    ConfigFeature,
+    MkosiArgs,
+    MkosiConfig,
+    OutputFormat,
+    QemuFirmware,
+    yes_no,
+)
+from mkosi.log import die
+from mkosi.partition import finalize_root, find_partitions
+from mkosi.qemu import (
+    KernelType,
+    copy_ephemeral,
+)
+from mkosi.run import (
+    read_subrange,
+    run,
+    spawn,
+)
+from mkosi.types import PathString
+from mkosi.util import INVOKING_USER
+
+
+def config_feature_to_tristate(cf: ConfigFeature) -> str:
+    if cf == ConfigFeature.enabled:
+        return "yes"
+    if cf == ConfigFeature.disabled:
+        return "no"
+    return ""
+
+
+def run_vmspawn(args: MkosiArgs, config: MkosiConfig) -> None:
+    if config.output_format not in (OutputFormat.disk, OutputFormat.directory):
+        die(f"{config.output_format} images cannot be booted in systemd-vmspawn")
+
+    if config.qemu_firmware == QemuFirmware.bios:
+        die("systemd-vmspawn cannot boot BIOS firmware images")
+
+    if config.qemu_cdrom:
+        die("systemd-vmspawn does not support CD-ROM images")
+
+    if config.tools_tree is not None:
+        die("systemd-vmspawn cannot be run with a tools tree")
+
+    if config.qemu_kernel:
+        kernel = config.qemu_kernel
+    else:
+        kernel = None
+
+    if kernel and not kernel.exists():
+        die(f"Kernel not found at {kernel}")
+
+    if config.qemu_firmware == QemuFirmware.auto:
+        if kernel:
+            firmware = QemuFirmware.uefi if KernelType.identify(kernel) != KernelType.unknown else QemuFirmware.linux
+        elif config.output_format == OutputFormat.directory or config.architecture.to_efi() is None:
+            firmware = QemuFirmware.linux
+        else:
+            firmware = QemuFirmware.uefi
+    else:
+        firmware = config.qemu_firmware
+
+    if firmware == QemuFirmware.linux:
+        kernel = config.output_dir_or_cwd() / config.output_split_kernel
+        if not kernel.exists():
+            die(
+                f"Kernel or UKI not found at {kernel}, please install a kernel in the image ",
+                hint="or provide a --qemu-kernel argument to mkosi vmspawn"
+            )
+
+    cmdline: list[PathString] = [
+        "systemd-vmspawn",
+        f"--qemu-smp={config.qemu_smp}",
+        f"--qemu-mem={config.qemu_mem}",
+        f"--qemu-kvm={config_feature_to_tristate(config.qemu_kvm)}",
+        f"--qemu-vsock={config_feature_to_tristate(config.qemu_vsock)}",
+        f"--vtpm={config_feature_to_tristate(config.qemu_swtpm)}",
+        f"--secure-boot={yes_no(config.secure_boot)}",
```

Comment:
Please use two arguments everywhere instead of all the format strings

Reply:
Fixed in dc05caea.